### PR TITLE
refactor(core): use JoinSet instead of FuturesUnordered

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit migrates "deno_core" from using "FuturesUnordered" to
"tokio::task::JoinSet". This makes every op to be a separate Tokio task
and should unlock better utilization of kqueue/epoll.

There were two quirks added to this PR:
- because of the fact that "JoinSet" immediately polls spawn tasks,
op sanitizers can give false positives in some cases, this was alleviated
by polling event loop once before running a test with "deno test", which
gives canceled ops an opportunity to settle
- "JsRuntimeState::waker" was moved to "OpState::waker" so that FFI
API can still use threadsafe functions - without this change the registered
wakers were wrong as they would not wake up the whole "JsRuntime"
but the task associated with an op

### CODE REVIEW

File: core/runtime.rs [hunk 1]
```diff
@@ -2433,16 +2433,21 @@ impl JsRuntime {
       let mut args: SmallVec<[v8::Local<v8::Value>; 32]> =
         SmallVec::with_capacity(32);
 
-      while let Poll::Ready(Some(item)) =
-        context_state.pending_ops.poll_next_unpin(cx)
-      {
-        let (promise_id, op_id, mut resp) = item;
+      loop {
+        let mut join_fut = context_state.pending_ops.join_next();
```
Comment (mmastrac) [inline core/runtime.rs:2437]: Ugh. I wonder if we can convince tokio to make `poll_join_next` public. We wouldn't need to pin this block.

File: core/runtime.rs [hunk 2]
```diff
@@ -2433,16 +2433,21 @@ impl JsRuntime {
...
+        // SAFETY: we are sure that the future won't move while it's being polled
+        let mut pinned_join_fut = unsafe { Pin::new_unchecked(&mut join_fut) };
+        let Poll::Ready(Some(item)) = pinned_join_fut.poll_unpin(cx) else {
+          break;
+        };
```
Comment (mmastrac) [inline core/runtime.rs:2442]: This doesn't fix the crash, but doesn't require `unsafe`:

```suggestion
        let item = {
          let next = std::pin::pin!(context_state.pending_ops.join_next());
          let Poll::Ready(Some(item)) = next.poll(cx) else {
            break;
          };
          item
        };
```

**Comment by bartlomieju** [comment]
I fixed the test panic, unfortunately now `js_realm_gc` test is hot spinning at 100% CPU. @mmastrac could you take a look?

File: core/runtime.rs (hunk 1 above)
Comment (bartlomieju) [reply core/runtime.rs]: Opened https://github.com/tokio-rs/tokio/issues/5769

**Comment by mmastrac** [comment]
> I fixed the test panic, unfortunately now `js_realm_gc` test is hot spinning at 100% CPU. @mmastrac could you take a look?

We looked into this test failure and it's probably an issue, but it's limited to realms and probably just means we're leaking tasks. We're going to punt fixing this to a later PR.

File: cli/tools/test.rs [hunk 3]
```diff
@@ -1008,6 +1010,21 @@ pub async fn test_specifier(
...
+    // TODO(bartlomieju): this is a nasty (beautiful) hack, that was required
+    // when switching `JsRuntime` from `FuturesUnordered` to `JoinSet`. With
+    // `JoinSet` all pending ops are immediately polled and that caused a problem
+    // when some async ops were fired and canceled before running tests (giving
+    // false positives in the ops sanitizer). We should probably rewrite sanitizers
+    // to be done in Rust instead of in JS (40_testing.js).
+    {
+      // Poll event loop once, this will allow all ops that are already resolved,
+      // but haven't responded to settle.
+      let waker = noop_waker();
```
Comment (mmastrac) [inline cli/tools/test.rs:1023]: Ouch. Does this approach fix the `js_realm_gc` test?

File: core/ops.rs [hunk 4]
```diff
@@ -184,6 +186,7 @@ pub struct OpState {
   pub tracker: OpsTracker,
   pub last_fast_op_error: Option<AnyError>,
   pub(crate) gotham_state: GothamState,
+  pub waker: Arc<AtomicWaker>,
```
Comment (mmastrac) [inline core/ops.rs:189]: Don't we have a waker on the runtime state as well?

File: ext/ffi/callback.rs [hunk 5]
```diff
@@ -99,21 +100,20 @@ struct CallbackInfo {
...
+  pub waker: Arc<AtomicWaker>,
 }
 
 impl Future for CallbackInfo {
   type Output = ();
   fn poll(
-    mut self: Pin<&mut Self>,
-    cx: &mut std::task::Context<'_>,
+    self: Pin<&mut Self>,
+    _cx: &mut std::task::Context<'_>,
```
Comment (mmastrac) [inline ext/ffi/callback.rs:110]: This FFI code is pretty wild, but this seems more correct.

**Review by mmastrac — APPROVED** [review]
LGTM, but I think this shows we really need to redesign the core FFI/NAPI dispatch. That's a pretty big piece of work, however.

File: ext/ffi/callback.rs (hunk 5 above)
Comment (aapoalas) [reply ext/ffi/callback.rs:110]: Oy! I made it with love and around 0% understanding on what I need to do :D

File: cli/tools/test.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/tools/test.rs:1023]: I don't think so. I'll look into that in a follow up

**Pull request merged**